torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -3944,11 +3952,22 @@ def _test_ddp_hook_with_optimizer_parity(
                     # Register hook that runs allreduce + functional optimizer
                     # step.
                     allreduce_hook = default.allreduce_hook
-                    opt_hook_state = optimizer_overlap_hooks._OptimizerHookState(
-                        functional_optim_cls,
-                        *functional_optim_args,
-                        **functional_optim_kwargs,
-                    )
+                    if construct_from_functional:
+                        f_opt = functional_optim_cls(
+                            [],
+                            *functional_optim_args,
+                            **functional_optim_kwargs,
+                            _allow_empty_param_list=True
```

Comment:
Oh, I see that you set `_allow_empty_param_list` before calling the factory function. In my opinion this is an internal implementation detail that the user should not be aware of. Is it possible to enforce this condition later inside `from_functional_optim`?

Reply:
Yes, this is constructing the functional optimizer. Eventually this construction logic is moved into `as_functional_optim`, and called by i.e. L60 in this PR: https://github.com/pytorch/pytorch/pull/71605/files. 

I think by the end of this stack, `as_functional_optim` which sets `_allow_empty_param_list` is only called in construction of `_OverlappedOptimizer` which is called by DDP, so the end user is not expected to interface with it.

It would be a little tricky to move it to `OptimizerHookState.from_functional_optim`, because `_allow_empty_param_list` needs to be specified in the constructor. We could refactor `OptimizerHookState.from_functional_optim` to construct the entire functional optimizer, but I don't think either approach will make too much of a difference as neither API is intended to be public and both would be called from the same location in `_OverlappedOptimizer` anyways.